From fec94d803ff4ff89158b0afd824f0b99465b6154 Mon Sep 17 00:00:00 2001
From: Yu Qiaowei <cerf.yu@rock-chips.com>
Date: Mon, 9 Mar 2026 19:46:15 +0800
Subject: [PATCH 33/87] video: rockchip: rga3: mpi_commit: restore
 output_params in rotate 90/270

Change-Id: I0fbdb0721a5e0bd52881b9383997e142a7a85c24
Signed-off-by: Yu Qiaowei <cerf.yu@rock-chips.com>
(cherry picked from commit dd28f74697bdfa0bf00091a20516b45890795068)
---
 drivers/video/rockchip/rga3/rga_drv.c | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/drivers/video/rockchip/rga3/rga_drv.c b/drivers/video/rockchip/rga3/rga_drv.c
index 7272111769e7d..5388cf919043b 100644
--- a/drivers/video/rockchip/rga3/rga_drv.c
+++ b/drivers/video/rockchip/rga3/rga_drv.c
@@ -138,6 +138,7 @@ int rga_mpi_commit(struct rga_mpi_job_t *mpi_job)
 	struct rga_req *cached_cmd;
 	struct rga_req mpi_cmd;
 	unsigned long flags;
+	bool need_swap_act = false;
 
 	request_manager = rga_drvdata->pend_request_manager;
 
@@ -174,6 +175,12 @@ int rga_mpi_commit(struct rga_mpi_job_t *mpi_job)
 	cached_cmd = request->task_list;
 	memcpy(&mpi_cmd, cached_cmd, sizeof(mpi_cmd));
 
+	/* rotate 90/270, driver requires that act_w/h be swapped. */
+	if (((mpi_cmd.rotate_mode & 0xf) == 1) &&
+	    ((mpi_cmd.sina == 65536 && mpi_cmd.cosa == 0) ||
+	     (mpi_cmd.sina == -65536 && mpi_cmd.cosa == 0)))
+		need_swap_act = true;
+
 	spin_unlock_irqrestore(&request->lock, flags);
 
 	/* set channel info */
@@ -198,10 +205,7 @@ int rga_mpi_commit(struct rga_mpi_job_t *mpi_job)
 					 &mpi_cmd.dst,
 					 &cached_cmd->dst);
 
-		/* rotate 90/270 */
-		if (((mpi_cmd.rotate_mode & 0xf) == 1) &&
-		    ((mpi_cmd.sina == 65536 && mpi_cmd.cosa == 0) ||
-		     (mpi_cmd.sina == -65536 && mpi_cmd.cosa == 0))) {
+		if (need_swap_act) {
 			swap(mpi_cmd.dst.act_w, mpi_cmd.dst.act_h);
 
 			if (request->flags & RGA_CONTEXT_DST_CACHE_INFO)
@@ -272,6 +276,10 @@ int rga_mpi_commit(struct rga_mpi_job_t *mpi_job)
 		mpi_job->output->vir_h = mpi_cmd.dst.vir_h;
 		mpi_job->output->rd_mode = mpi_cmd.dst.rd_mode;
 		mpi_job->output->format = mpi_cmd.dst.format;
+
+		/* restore data that has been swapped */
+		if (need_swap_act)
+			swap(mpi_job->output->width, mpi_job->output->height);
 	}
 
 err_put_request:
